subsys/net/lib/sockets/getaddrinfo.c

```diff
@@ -346,6 +354,65 @@ static inline int z_vrfy_z_zsock_getaddrinfo_internal(const char *host,
 
 #endif /* defined(CONFIG_DNS_RESOLVER) */
 
+#if defined(CONFIG_NET_IPV6)
+static int try_resolve_literal_ipv6(const char *host, const char *service,
+				    const struct zsock_addrinfo *hints,
+				    struct zsock_addrinfo *res)
+{
+	int family = AF_UNSPEC;
+	long port = 0;
+	int ret;
+	bool result;
+	int socktype = SOCK_STREAM;
+	int protocol = IPPROTO_TCP;
+
+	if (hints) {
+		family = hints->ai_family;
+		if (hints->ai_socktype == SOCK_DGRAM) {
+			socktype = SOCK_DGRAM;
+			protocol = IPPROTO_UDP;
+		}
+	}
+
+	if (service) {
+		port = strtol(service, NULL, 10);
+		if (port < 1 || port > 65535) {
+			return DNS_EAI_NONAME;
+		}
+	}
+
+	if (host == NULL) {
+		/* Cannot parse literal IPv6 address if not given. */
+		return DNS_EAI_NONAME;
+	}
+
+	if (family == AF_INET) {
+		/* Cannot parse IPv4 */
+		return DNS_EAI_NONAME;
+	}
+
+	result = net_ipaddr_parse(host, strlen(host), &res->_ai_addr);
+
+	if (result && (net_sin6(&res->_ai_addr)->sin6_family == AF_INET6)) {
```

Comment:
> You don't know at this point whether to use IPv4/IPv6 structure, it would be more appropriate to check against generic `sockaddr`, i. e. `res->_ai_addr.sa_family`

The sa_family fields in different structs point to same location (first two bytes of those structs), so it does not matter which family field we are using.

Reply:
Yes, I understand that this won't make a difference to the compiler, I was more thinking of folks analyzing this code in the future - one may wonder why is IPv6 assumed before it's actually verified that we deal with IPv6.

Reply:
Sure, it makes it more clear to the user if the generic sockaddr struct family value is used